x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskTests.java

```diff
@@ -5,205 +5,623 @@
-        });
-    }
-
-    public void testMappingUpdate() throws Exception {
-        task = createShardFollowTask(1024, 1, 1, 1000, 1024, -1);
-        task.start(-1);
-
-        assertBusy(() -> {
-            assertThat(task.getStatus().getFollowerGlobalCheckpoint(), greaterThanOrEqualTo(1000L));
-        });
-        imdVersion.set(2L);
-        leaderGlobalCheckPoint.set(10000L);
-        assertBusy(() -> {
-            assertThat(task.getStatus().getFollowerGlobalCheckpoint(), equalTo(10000L));
-        });
-        assertThat(mappingUpdateCounter.get(), equalTo(2));
-    }
-
-    public void testOccasionalApiFailure() throws Exception {
-        task = createShardFollowTask(1024, 1, 1, 10000, 1024, -1);
-        task.start(-1);
-        randomlyFailWithRetryableError.set(true);
-        assertBusy(() -> {
-            assertThat(task.getStatus().getFollowerGlobalCheckpoint(), equalTo(10000L));
-        });
-        assertThat(failedRequests.get(), greaterThan(0));
-    }
-
-    public void testNotAllExpectedOpsReturned() throws Exception {
-        task = createShardFollowTask(1024, 1, 1, 10000, 1024, -1);
-        task.start(-1);
-        randomlyTruncateRequests.set(true);
-        assertBusy(() -> {
-            assertThat(task.getStatus().getFollowerGlobalCheckpoint(), equalTo(10000L));
-        });
-        assertThat(truncatedRequests.get(), greaterThan(0));
-    }
-
-    ShardFollowNodeTask createShardFollowTask(int maxBatchOperationCount, int maxConcurrentReadBatches, int maxConcurrentWriteBathces,
-                                              int globalCheckpoint, int bufferWriteLimit, long followGlobalCheckpoint) {
-        leaderGlobalCheckPoint = new AtomicLong(globalCheckpoint);
-        imdVersion = new AtomicLong(1L);
-        mappingUpdateCounter = new AtomicInteger(0);
-        randomlyTruncateRequests = new AtomicBoolean(false);
-        truncatedRequests = new AtomicInteger();
-        randomlyFailWithRetryableError = new AtomicBoolean(false);
-        failedRequests = new AtomicInteger(0);
+    private Exception fatalError;
+    private List<long[]> shardChangesRequests;
+    private List<List<Translog.Operation>> bulkShardOperationRequests;
+    private BiConsumer<TimeValue, Runnable> scheduler = (delay, task) -> task.run();
+
+    private Queue<Exception> readFailures;
+    private Queue<Exception> writeFailures;
+    private Queue<Exception> mappingUpdateFailures;
+    private Queue<Long> imdVersions;
+    private Queue<Long> followerGlobalCheckpoints;
+
+    public void testCoordinateReads() {
+        ShardFollowNodeTask task = createShardFollowTask(8, 8, 8, Integer.MAX_VALUE, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(8));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(8L));
```

Comment:
Can we fold all these assertions into a single one? I think this should cover enough.

```
   assertThat(shardChangesRequests, contains(new long[][]{
            {0L, 8L}, {9L, 8L}, {18L, 8L}, {27L, 8L}, {36L, 8L}, {45L, 8L}, {54L, 8L}, {63L, 8L}
        }));
```

Moreover, the leader should not return more than the requesting batch size. Here, we request 8 operations, but it returns 9 operations.

Reply:
That is not the end offset but the `maxBatchOperationCount`: https://github.com/elastic/elasticsearch/pull/32121/files#diff-e143edbe5e6171625fe27ec29039aabdR622

Since that is always the same perhaps it not worth asserting it. Should we just assert the from offset?